New service that waits on zvol links to be created

pzakha opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

This creates a new service, `zfs-volume.service`. When this service becomes active, it ensures that all the zvols of imported pools are ready to use and their symlinks are created under `/dev`. Note that this service doesn't actually do any modifications to the system, it only waits for resources to be ready before returning. Any service that performs operations on zvols at boot time would want to add this service as a dependency.

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
The `rtslib-fb-targetctl` service that comes with the `python-rtslib-fb` package provides a userland interface to the LIO framework, which can serve iSCSI LUNs from block devices on the system. ZFS volumes (zvols) can be used as backing block devices. However, there is currently no guarantee that the zvols are ready when that service runs at boot time. If that is the case, then `rtslib-fb-targetctl` will fail to create LUNs from those zvols. The new `zfs-volume` service can be added as a dependency to `rtslib-fb-targetctl` to avoid such situations.

### Description
<!--- Describe your changes in detail -->
The original Delphix review is here: https://github.com/delphix/zfs/pull/56.
The idea is to list all the zvols and then wait for their links under `/dev` to be created.

### How Has This Been Tested?
- See https://github.com/delphix/zfs/pull/56 for original testing.
- Made sure that the `rtslib-fb-targetctl` service doesn't report any errors at boot because of unexisting zvol links after adding a dependency on zfs-volume.service.
- Tested the new script converted to POSIX sh:
1. Reduced the inner loop iterations to test that logic better:
```
Jun 28 22:11:57 localhost systemd[1]: Starting Wait for ZFS Volume (zvol) /dev links...
Jun 28 22:12:03 localhost zvol-wait[4899]: Testing 1003 zvol links
Jun 28 22:12:04 localhost zvol-wait[4899]: Still waiting on 120 zvol links ...
Jun 28 22:12:04 localhost zvol-wait[4899]: All zvol links are now present.
Jun 28 22:12:04 localhost systemd[1]: Started Wait for ZFS Volume (zvol) /dev links.
```
2. Simulated a zvol deletion while the service is running with additional test code:
```
-- Logs begin at Wed 2019-06-19 14:49:06 UTC, end at Fri 2019-06-28 22:19:04 UTC. --
Jun 28 22:17:59 localhost systemd[1]: Starting Wait for ZFS Volume (zvol) /dev links...
Jun 28 22:18:07 pzakha-lx-4.dcenter zvol-wait[3998]: Testing 1003 zvol links
Jun 28 22:18:07 pzakha-lx-4.dcenter zvol-wait[3998]: TEST CODE: delete domain0/VOLS/zvol333
Jun 28 22:18:13 pzakha-lx-4.dcenter zvol-wait[3998]: Still waiting on 1 zvol links ...
Jun 28 22:18:18 pzakha-lx-4.dcenter zvol-wait[3998]: Still waiting on 1 zvol links ...
Jun 28 22:18:18 pzakha-lx-4.dcenter zvol-wait[3998]: No progress since last loop.
Jun 28 22:18:18 pzakha-lx-4.dcenter zvol-wait[3998]: Checking if any zvols were deleted.
Jun 28 22:18:18 pzakha-lx-4.dcenter zvol-wait[3998]: 1 zvol(s) deleted.
Jun 28 22:18:18 pzakha-lx-4.dcenter zvol-wait[3998]: All zvol links are now present.
Jun 28 22:18:18 pzakha-lx-4.dcenter systemd[1]: Started Wait for ZFS Volume (zvol) /dev links.
```

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

pzakha commented:
cc: @behlendorf @grwilson @jgallag88 @prakashsurya

behlendorf commented on debian/tree/zfsutils-linux/usr/lib/zfs-linux/zvol-wait:
```diff
@@ -0,0 +1,84 @@
+#!/bin/sh
```
I'd suggest moving this script in to the `etc/systemd/system/` directory, or a new `etc/systemd/scripts` directory in the git tree.  The right install location for this on the system appears to be `/usr/lib/systemd/scripts/` (checked on CentOS 7 and Ubuntu 18.04), but it should also be possible to override it with `./configure`.  You'll  need to add a `systemdscriptsdir` to `config/user-systemd.m4` like the others for this.

pzakha replied:
Looking at other systemd ExecStart scripts, most of the system ones are placed into `/lib/systemd/` directly and are named `systemd-something`.

A bunch of other scripts are placed all around the place, with many in `/bin/`, `/usr/bin/`, `/sbin`, etc.

Thinking about it a bit more, it might be convenient to have this function available for manual execution, so having it in `/bin` or `/usr/bin` seems to make more sense to me. We could rename it to something like `zfs-volume-wait`.

behlendorf replied:
I wasn't able to dig up the documentation to confirm this, but it looks to me like everything in `/lib/systemd/` in an arch specific binary, and `/lib/systemd/scripts/` is for portable arch-independent scripts.

But I agree, if we want to be able to easily use this utility outside of systemd then it really belongs in `/usr/bin/`.  In which case I'd suggest putting the source in a subdirectory under ./cmd/, which is where we put things like `zvol_id`.

behlendorf commented on etc/systemd/system/Makefile.am:
```diff
@@ -8,6 +8,7 @@ systemdunit_DATA = \
 	zfs-mount.service \
 	zfs-share.service \
 	zfs-import.target \
+	zfs-volume.service \
```
nit: since there already aren't alphabetic I'd move this up below the other .service files

behlendorf commented on etc/systemd/system/zfs-volume.service.in:
```diff
@@ -0,0 +1,13 @@
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+ExecStart=/usr/lib/zfs-linux/zvol-wait
```
You might not need to specify a full path here if it's installed in `/usr/lib/systemd/scripts/`, otherwise you'll need to substitute `@systemdscriptsdir@` and update the `*.in` rule in the `Makefile.am` to do the substitution.

jgallag88 commented on cmd/zfs_volume_wait/zfs_volume_wait:
```diff
@@ -0,0 +1,84 @@
+
+filter_out_zvols_with_links() {
+	while read -r zvol; do
+		if [ ! -e "/dev/$zvol" ]; then
```
Should check `/dev/zvol/$zvol` instead of `/dev/$zvol`? According to the comment in `60-zvol.rules.in`, the `/dev/$zvol` link exists only for compatibility, and I'm guessing it could be missing if there is a collision between a pool name and some other file in `/dev/`

grwilson replied:
You may also want to use `-L` instead of `-e` to ensure if it's a link incase someone drops a file in this directory.

pzakha replied:
done

jgallag88 commented on man/man1/zfs_volume_wait.1:
```diff
@@ -0,0 +1,19 @@
+When a ZFS pool is imported, ZFS will register each ZFS volume
+(zvol) as a disk device with the system. As the disks are registered,
+.Xr \fBudev 7\fR
+will asynchronously create symlinks under /dev using the zvol's name.
```
[nit] We might want to change `/dev` to be `/dev/zvol` specifically, since that seems to be the preferred link.

behlendorf commented on cmd/Makefile.am:
```diff
@@ -5,4 +5,4 @@ if USING_PYTHON
 endif
 
-SUBDIRS += mount_zfs zed zvol_id
+SUBDIRS += mount_zfs zed zvol_id zfs_volume_wait
```
To resolve this build error you'll need to add the new utility to the spec file.

```
Checking for unpackaged file(s): /usr/lib/rpm/check-files /tmp/zfs-build-buildbot-XOz43uTI/BUILDROOT/zfs-0.8.0-106_g663bba6.el7.x86_64
error: Installed (but unpackaged) file(s) found:
   /usr/bin/zfs_volume_wait
```

This should work.

```diff
index 0b16cd0..989a31b 100644
--- a/rpm/generic/zfs.spec.in
+++ b/rpm/generic/zfs.spec.in
@@ -418,6 +418,7 @@ systemctl --system daemon-reload >/dev/null || true
 %{_sbindir}/*
 %{_bindir}/raidz_test
 %{_bindir}/zgenhostid
+%{_bindir}/zfs_volume_wait
 # Optional Python 2/3 scripts
 %{_bindir}/arc_summary
 %{_bindir}/arcstat
```

Would it be more convenient and consistent to call the command `zvol_wait` similar to `zvol_id`?  Or do we want the longer name, I don't have a preference but wanted to ask.

pzakha replied:
> Would it be more convenient and consistent to call the command zvol_wait similar to zvol_id? Or do we want the longer name, I don't have a preference but wanted to ask.

I'm fine either way. I've called it zfs_volume_wait because the service was renamed to `zfs-volume.service`.

pzakha replied:
renamed the script to `zvol_wait`

rlaager commented:
Some random questions. Note that these really _are_ questions, and I'm _not_ implying that I'm in favor of them.

1. Should this be zfs-volume.target instead (or, more likely, should there be a zfs-volume.target that is after and/or requires a zfs-volume-wait.service)?
2. Did you consider a template unit that waits for one particular zvol?
3. Did you consider a generator (probably with a cache, like we use for filesystem mounts), possibly in combination with a template unit?

@aerusso @Fabian-Gruenbichler do you have any thoughts here?

Fabian-Gruenbichler commented:
@rlaager I had many of the same thoughts/questions, but not yet found the time to write them up.

Especially in combination with encryption it would be nice to have individual units for zvols (that could then also be ordered after/depend on the resp. zfs-load-key units, if necessary).

I am also not sure whether we really want to tie in this new service(s)/target into the main zfs.target. IMHO it would make more sense for other services (like iSCSI targets, hypervisors that have guest volumes as zvols, ...) that actually depend on the links being there to explicitly pull it/them in. This would avoid artificially slowing down the default boot - there were no guarantees so far that all zvol links are available, so nobody should rely on this (yet).

aerusso commented:
@behlendorf At the risk of seeing everything as a nail (because of the hammer that I've wielded), I agree that the granularity of having `zfs-volume@.service` looks appealing. You might also be able to use `.path` units instead of shellscript to check for zvol readiness.

I have not tested the below, but this is roughly what I'm thinking:

https://github.com/aerusso/zfs/commit/ce2abd448bfabfbcd3a6eff3af21ef3b13ce8d36 in https://github.com/aerusso/zfs/commits/pulls/systemd-zvols

behlendorf commented:
> the granularity of having zfs-volume@.service looks appealing. You might also be able to use .path units instead of shellscript to check for zvol readiness.

The granularity would be nice, but I'd be nice to not have that additional requirement block this PR.  I'm not familiar with .path units would this be something which could co-exist with the proposed approach?  Similar to what's done with `zfs-mount` service and generator?

I do think we want to have a `zfs-volume.target` and have it require a `zfs-volume-wait.service`.

pzakha commented:
Thanks for the feedback. I do think that having more granular units could be very useful in some cases, but even with the implementation stub from @aerusso we would probably need to spend much more time testing it and ironing out any corner cases. If others do not object, I'd prefer going forward with this approach for now since it does provide quite some value already. With `zfs-volume.target` in place, we can later have it target a more granular set of units.

pzakha commented:
> This would avoid artificially slowing down the default boot - there were no guarantees so far that all zvol links are available, so nobody should rely on this (yet).

For now I will not add any `WantedBy` for `zfs-volume.target`, so if no one needs it, it will not incur the extra `zfs list` call from `zfs-volume-wait.service`

pzakha commented:
I have added the new `zfs-volume.target` unit and tested that it is not being run by default unless it is specifically `Wanted` or `Required` by another service.

behlendorf commented on etc/systemd/system/50-zfs.preset.in:
```diff
@@ -5,4 +5,5 @@ enable zfs-import.target
 enable zfs-mount.service
 enable zfs-share.service
 enable zfs-zed.service
+enable zfs-volume-wait.service
```
I thought we determined this service should be disabled by default for now.

pzakha replied:
Since `zfs-volume.target` is disabled by default, this service will not be started by anyone even though it is enabled. It will be started only if `zfs-volume.target` is `WantedBy` or `RequiredBy` another service.  I'm not sure if that's a common way to do those things in systemd, but with this configuration `zfs-volume-wait.service` would not start on boot by default.

behlendorf replied:
I see, that makes sense.  I couldn't find documentation for the established convention either and that seems reasonable.

loli10K commented:
This implementation does not seem to handle `volmode=none`:

```
root@linux:/usr/src/zfs# zfs create -p $POOLNAME/backups/path/to/zvol -V 1G -s
root@linux:/usr/src/zfs# 
root@linux:/usr/src/zfs# time cmd/zvol_wait/zvol_wait
Testing 1 zvol links
All zvol links are now present.

real	0m0.008s
user	0m0.000s
sys	0m0.000s
root@linux:/usr/src/zfs# 
root@linux:/usr/src/zfs# zfs set volmode=none $POOLNAME/backups/path/to/zvol
root@linux:/usr/src/zfs# 
root@linux:/usr/src/zfs# time cmd/zvol_wait/zvol_wait
Testing 1 zvol links
Still waiting on 1 zvol links ...
No progress since last loop.
Checking if any zvols were deleted.

< more logging >

Still waiting on 1 zvol links ...
No progress since last loop.
Checking if any zvols were deleted.
Remaining zvols:
testpool/backups/path/to/zvol
Timed out waiting on zvol links

real	10m1.814s
user	0m0.052s
sys	0m0.296s
root@linux:/usr/src/zfs# 
```

Systems with at least 1 "hidden" (`volmode=none`) ZVOL should not have to wait 10m, especially if the ZVOL is not expected to show up by user choice.

pzakha commented:
> This implementation does not seem to handle `volmode=none`

Thanks for the review. I was not aware of this property. I'll update the review to filter out zvols with `volmode=none`.

pzakha commented:
Updated code with the following changes:
- Account for `volmode=none`
- Rename `zfs-volume.target` -> `zfs-volumes.target`
- Added `WantedBy=zfs.target` to `zfs-volumes.target` so that the target is reached on boot when enabled.

behlendorf commented:
Thanks for addressing those last few issues.  I'll get this merged pending the final CI results.

behlendorf commented:
@pzakha bad timing, I believe we had some zimport failures due to the unrelated log space map changes being integrated.  Can you rebase this one last time on master.
